certora/specs/SocialRecoveryModule.spec

```
@@ -1,26 +1,212 @@
+function requireGuardiansLinkedListIntegrity(address guardian) {
+    uint256 index;
+    require index < guardianStorageContract.countGuardians(safeContract);
+    require currentContract.isGuardian(safeContract, guardian) =>
+        guardianStorageContract.getGuardians(safeContract)[index] == guardian;
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+}
+
+// This integrity rule verifies that if the addGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the new guardian is added to the guardian list,
+// - and no other account (guardian or not) is affected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherAccount) {
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold) {
+    requireSocialRecoveryModuleEnabled();
+
+    // No value should be sent with the transaction.
+    require e.msg.value == 0;
+    // The threshold should be greater than 0. We can't have guardian(s) with a threshold of 0.
+    require threshold > 0;
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;    
+    // The guardian count should be less than the maximum value to prevent overflow.
+    require currentGuardiansCount < max_uint256; // To prevent overflow (Realistically can't reach).
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    // The guardian should not be values such as zero, sentinel, or the Safe contract itself.
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+
+    // The guardian should not be an owner of the Safe contract at the time of addition.
+    require !safeContract.isOwner(guardian);
+    // The guardian should not be already added as a guardian.
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    // The threshold should be less than or equal to the total number of guardians.
+    require threshold <= guardianStorageContract.countGuardians(safeContract);
```

Actually, I think this is wrong. Specifically, if I have 1 guardian, I should be able to use a threshold of 2 when adding a new guardian (since I will have enough after the adding).

---

Also, I think the harness function is needed for linked list integrity, but I do believe we should be using `count` as much as possible:

```suggestion
    require threshold <= guardianStorageContract.entries[safeContract].count + 1;
```

---

Updated 👍🏾